key_management/src/keystore.rs

```diff
@@ -113,127 +117,387 @@ pub trait KeyStore {
         self.key_info.get(k).cloned().ok_or(Error::KeyInfo)
     }
 
-    fn put(&mut self, key: String, key_info: KeyInfo) -> Result<(), Error> {
+    /// Save a key key_info pair to the KeyStore
+    pub fn put(&mut self, key: String, key_info: KeyInfo) -> Result<(), Error> {
         if self.key_info.contains_key(&key) {
             return Err(Error::KeyExists);
         }
         self.key_info.insert(key, key_info);
-        let file = OpenOptions::new()
-            .write(true)
-            .create(true)
-            .open(&self.location)
-            .map_err(|err| Error::Other(err.to_string()))?;
-        serde_json::to_writer(&file, &self.key_info)
-            .map_err(|err| Error::Other(err.to_string()))?;
+
+        if self.persistence.is_some() {
+            self.flush()?;
+        }
+
         Ok(())
     }
 
-    fn remove(&mut self, key: String) -> Result<KeyInfo, Error> {
+    /// Remove the Key and corresponding key_info from the KeyStore
+    pub fn remove(&mut self, key: String) -> Result<KeyInfo, Error> {
         let key_out = self.key_info.remove(&key).ok_or(Error::KeyInfo)?;
-        let file = OpenOptions::new()
-            .write(true)
-            .create(true)
-            .open(&self.location)
-            .map_err(|err| Error::Other(err.to_string()))?;
-        serde_json::to_writer(file, &self.key_info).map_err(|err| Error::Other(err.to_string()))?;
+
+        if self.persistence.is_some() {
+            self.flush()?;
+        }
+
         Ok(key_out)
     }
 }
+
+impl EncryptedKeyStore {
+    fn derive_key(
+        passphrase: &str,
+        prev_salt: Option<Vec<u8>>,
+    ) -> Result<(pwhash::Salt, Arc<secretbox::Key>), EncryptedKeyStoreError> {
+        let salt = match prev_salt {
+            Some(prev_salt) => match pwhash::Salt::from_slice(&prev_salt) {
+                Some(salt) => salt,
+                None => return Err(EncryptedKeyStoreError::ConfigurationError),
+            },
+            None => pwhash::gen_salt(),
+        };
+        let mut key = secretbox::Key([0; secretbox::KEYBYTES]);
+
+        let secretbox::Key(ref mut kb) = key;
+        pwhash::derive_key(
+            kb,
+            passphrase.as_bytes(),
+            &salt,
+            pwhash::OPSLIMIT_INTERACTIVE,
+            pwhash::MEMLIMIT_INTERACTIVE,
+        )
+        .unwrap();
+
+        Ok((salt, Arc::new(key)))
+    }
+
+    fn encrypt(encryption_key: Arc<secretbox::Key>, msg: &[u8]) -> Vec<u8> {
+        let nonce = secretbox::gen_nonce();
+
+        let mut ciphertext = secretbox::seal(msg, &nonce, &encryption_key);
+        ciphertext.append(&mut nonce.as_ref().to_vec());
+        ciphertext
+    }
+
+    fn decrypt(
+        encryption_key: Arc<secretbox::Key>,
+        msg: &[u8],
+    ) -> Result<Vec<u8>, EncryptedKeyStoreError> {
+        let ciphertext = &msg[..msg.len() - 24];
+
+        let nonce = match secretbox::Nonce::from_slice(&msg[msg.len() - 24..]) {
```

Comment:
`ok_or` is perfect for these situations where we want to return a Result in None case.

Reply:
This makes it much cleaner (cause `?` op). Thanks for this.

Reply:
Yeah, i love how std lib APIs compose well together